axelrod/result_set.py

```diff
@@ -28,314 +33,297 @@ def wrapper(*args):
+                for repetition in range(self.repetitions):
+                    key = (repetition, player_index, opponent_index)
+                    if key in payoffs_dict:
+                        row.append(payoffs_dict[key])
+                matrix.append(row)
+            payoffs.append(matrix)
+        return payoffs
 
     @update_progress_bar
-    def _build_vengeful_cooperation(self):
+    def _build_score_diffs(self, payoff_diffs_series):
         """
-        Returns:
-        --------
+        Parameters
+        ----------
 
-            The vengeful cooperation matrix derived from the
-            normalised cooperation matrix:
+            payoffs_diffs_series : pandas.Series
 
-                Dij = 2(Cij - 0.5)
-        """
-        return [[2 * (element - 0.5) for element in row]
-                for row in self.normalised_cooperation]
-
-    @update_progress_bar
-    def _build_payoff_diffs_means(self):
-        """
         Returns:
         --------
-
-            The score differences between players.
+            The mean of per turn payoff differences
             List of the form:
-
             [ML1, ML2, ML3..., MLn]
-
             Where n is the number of players and MLi is a list of the form:
-
             [pi1, pi2, pi3, ..., pim]
-
-            Where pij is the mean difference of the
-            scores per turn between player i and j in repetition m.
-        """
-        payoff_diffs_means = [[mean(diff) for diff in player]
-                              for player in self.score_diffs]
-        return payoff_diffs_means
-
-    @update_progress_bar
-    def _build_payoff_stddevs(self):
+            Where m is the number of players and pij is a list of the form:
+            [uij1, uij2, ..., uijk]
+            Where k is the number of repetitions and u is the mean utility
+            difference (over all repetitions) obtained by player i against
+            player j.
         """
-        Returns:
-        --------
-
-            The mean of per turn payoffs.
-            List of the form:
 
-            [ML1, ML2, ML3..., MLn]
-
-            Where n is the number of players and MLi is a list of the form:
+        payoff_diffs_dict = payoff_diffs_series.to_dict()
+        score_diffs = []
+        for player_index in range(self.num_players):
+            matrix = []
+            for opponent_index in range(self.num_players):
+                row = []
+                for repetition in range(self.repetitions):
+                    row.append(payoff_diffs_dict.get((repetition,
+                                                      player_index,
+                                                      opponent_index,
+                                                      ), 0))
+                matrix.append(row)
+            score_diffs.append(matrix)
+        return score_diffs
 
-            [pi1, pi2, pi3, ..., pim]
+    @update_progress_bar
+    def _build_match_lengths(self, length_series):
```

Comment:
Seems like these functions could be abstracted, they are all very similar.

Reply:
I thought that and started taking a look at it but they've all got tiny differences that made it a bit painful. I'll take another look :) :+1: